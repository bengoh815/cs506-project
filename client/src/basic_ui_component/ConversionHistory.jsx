import React from "react";
import "./ConversionHistory.css";
import { useEffect, useState } from "react";

// Mock data for conversion history
const mockConversionHistoryData = [
  { fileName: "song1.mid", author: "Alice", date: "2024-03-10" },
  { fileName: "tune2.mid", author: "Bob", date: "2024-03-11" },
  { fileName: "melody3.mid", author: "Charlie", date: "2024-03-12" },
];

const ConversionHistory = () => {
  const [convertedFiles, setConvertedFiles] = useState([]);

  useEffect(() => {
    // Fetch the conversion history from the backend
    fetch("some/url", {
      method: "GET",
      headers: {
        // TODO: Add headers
      },
    })
      .then((response) => {
        console.log("Fetching conversion history from the backend...");
        // return response.json();
      })
      .then((data) => {
        setConvertedFiles(data);
        console.log(convertedFiles);
      })
      .catch((error) => {
        console.error("Error:", error);
      });
  }, []);

  return (
    <div className="conversion-history-container">
      <div id="conversion-history">
        <h2 id="conversion-history-heading">Conversion History</h2>
        <table id="conversion-history-table">
          <thead>
            <tr>
              <th>File</th>
              <th>Author</th>
              <th>Date</th>
            </tr>
          </thead>
          <tbody>
            {mockConversionHistoryData.map((entry, index) => (
              <tr key={index}>
                <td>{entry.fileName}</td>
                <td>{entry.author}</td>
                <td>{entry.date}</td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    </div>
  );
};

export default ConversionHistory;
