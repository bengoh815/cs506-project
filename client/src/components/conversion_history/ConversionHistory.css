.conversion-history-container {
  display: flex;
  justify-content: center;
  align-items: center;
}

#conversion-history-table {
  border-collapse: collapse;
  width: 100%;
  margin: 0 auto;
}

#conversion-history-table th,
#conversion-history-table td {
  border: 2px solid black;
  padding: 8px;
}

#conversion-history-table thead {
  background-color: #f2f2f2;
}

input[type="text"] {
  width: 100%;
  padding: 10px;
  margin: 10px 0;
  box-sizing: border-box;
  border: 2px solid #ddd;
  border-radius: 4px;
}

#conversion-history-table th {
  height: 50px;
}

.pagination {
  list-style: none;
  padding: 0;
  display: flex;
  justify-content: center;
}

.pagination li {
  margin-top: 20px;
}

.pagination li a {
  cursor: pointer;
  border: 1px solid #ddd;
  padding: 5px 10px;
  text-decoration: none;
  color: #333;
  margin: 5px;
}

.pagination li a:hover,
.pagination li a.active {
  background-color: #eee;
}

.pagination__link--active a {
  background-color: #007bff;
  color: white;
}

.pagination__link--disabled a {
  color: #ccc;
}

.pagination__link--disabled {
  color: #ccc;
  cursor: not-allowed;
  pointer-events: none;
}

.pagination__link--disabled a {
  border-color: #ccc;
}

.sortable:hover {
  cursor: pointer;
}

#conversion-history-table td button {
  padding: 5px 10px;
  background-color: #007bff;
  color: white;
  border: none;
  cursor: pointer;
  margin-right: 5px;
  border-radius: 10px;
}

#conversion-history-table td button:hover {
  background-color: #0056b3;
}