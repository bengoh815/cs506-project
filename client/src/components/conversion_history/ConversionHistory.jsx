/******************************************************************************
 * Filename: ConversionHistory.jsx
 * Purpose:  A component that displays the user's file conversion history.
 * Author:   Victor Nguyen & Don Ma
 *
 * Description:
 * This file contains the ConversionHistory component that displays the user's
 * file conversion history. The component fetches the user's conversion history
 * from the backend and displays it in a table. The table contains the file name,
 * author, and date of each conversion. The component is designed to be displayed
 * as part of the main application content.
 *
 * Usage:
 * To use this component, simply import it into the desired file and render it.
 *
 * Notes:
 * [Any additional notes, considerations, or important information
 * about the file that may be relevant to developers or users.]
 *
 ******************************************************************************/

import React, { useEffect, useState } from "react";
import "./ConversionHistory.css";
import ReactPaginate from "react-paginate";

// Mock data for conversion history
const mockConversionHistoryData = [
  { 'fileName': 'song1.mid', 'author': 'Fiona', 'date': '2024-03-10', 'size': '3.4 MB' },
  { 'fileName': 'song2.mid', 'author': 'George', 'date': '2024-03-11', 'size': '3.5 MB' },
  { 'fileName': 'song3.mid', 'author': 'George', 'date': '2024-03-12', 'size': '5.4 MB' },
  { 'fileName': 'song4.mid', 'author': 'Hannah', 'date': '2024-03-13', 'size': '35.41 MB' },
  { 'fileName': 'song5.mid', 'author': 'Julia', 'date': '2024-03-14', 'size': '9.4 MB' },
  { 'fileName': 'song6.mid', 'author': 'Ethan', 'date': '2024-03-15', 'size': '13.0 MB' },
  { 'fileName': 'song7.mid', 'author': 'George', 'date': '2024-03-16', 'size': '31.5 MB' },
  { 'fileName': 'song8.mid', 'author': 'George', 'date': '2024-03-17', 'size': '3.4 MB' },
  { 'fileName': 'song9.mid', 'author': 'Fiona', 'date': '2024-03-18', 'size': '3.9 MB' },
  { 'fileName': 'song10.mid', 'author': 'George', 'date': '2024-03-19', 'size': '1.4 MB' },
  { 'fileName': 'song11.mid', 'author': 'Diana', 'date': '2024-03-20', 'size': '254 MB' },
  { 'fileName': 'song12.mid', 'author': 'Alice', 'date': '2024-03-21', 'size': '3.0 MB' },
  { 'fileName': 'song13.mid', 'author': 'Bob', 'date': '2024-03-22', 'size': '0.9 MB' },
];

const ConversionHistory = () => {
  const [convertedFiles, setConvertedFiles] = useState([]);
  const [currentPage, setCurrentPage] = useState(0);
  const itemsPerPage = 10;
  const [sortingCriteria, setSortingCriteria] = useState('fileName'); // default is fileName
  const [isAscending, setIsAscending] = useState(true);

  useEffect(() => {
    fetch("http://localhost:8765/api/v1/midis", {
      method: "GET",
      headers: {

      },
    })
      .then((response) => {
        console.log("Fetching conversion history from the backend...");
      })
      .then((data) => {
        setConvertedFiles(data);
        console.log(convertedFiles);
      })
      .catch((error) => {
        console.error("Error:", error);
      });
  }, []);

  // Sorting function
  const sortItems = (a, b) => {
    if (sortingCriteria === 'fileName') {
      return isAscending ? a.fileName.localeCompare(b.fileName) : b.fileName.localeCompare(a.fileName);
    } else if (sortingCriteria === 'author') {
      return isAscending ? a.author.localeCompare(b.author) : b.author.localeCompare(a.author);
    } else if (sortingCriteria === 'date') {
      return isAscending ? new Date(a.date) - new Date(b.date) : new Date(b.date) - new Date(a.date);
    } else if (sortingCriteria === 'size') {
      const sizeA = parseFloat(a.size);
      const sizeB = parseFloat(b.size);
      return isAscending ? sizeA - sizeB : sizeB - sizeA;
    }
  };

  mockConversionHistoryData.sort(sortItems);

  // Calculate the current items to display
  const indexOfLastItem = (currentPage + 1) * itemsPerPage;
  const indexOfFirstItem = indexOfLastItem - itemsPerPage;
  const currentItems = mockConversionHistoryData.slice(indexOfFirstItem, indexOfLastItem);


  // Change page handler for ReactPaginate
  const changePage = ({ selected }) => {
    setCurrentPage(selected);
  };

  // Handler to update sorting criteria and direction
  const handleSort = (newCriteria) => {
    if (sortingCriteria === newCriteria) {
      setIsAscending(!isAscending);
    } else {
      setSortingCriteria(newCriteria);
      setIsAscending(true);
    }
  };

  const getSortingIndicator = (columnName) => {
    if (sortingCriteria === columnName) {
      return isAscending ? ' ▲' : ' ▼';
    }
    return '';
  };

  return (
    <div className="conversion-history-container">
      <div id="conversion-history">
        <h2 id="conversion-history-heading">Conversion History</h2>
        <table id="conversion-history-table">
          <thead>
            <tr>
              <th onClick={() => handleSort('fileName')} className="sortable">
                File{getSortingIndicator('fileName')}
              </th>
              <th onClick={() => handleSort('author')} className="sortable">
                Author{getSortingIndicator('author')}
              </th>
              <th onClick={() => handleSort('date')} className="sortable">
                Date{getSortingIndicator('date')}
              </th>
              <th onClick={() => handleSort('size')} className="sortable">
                Size{getSortingIndicator('size')}
              </th>
            </tr>
          </thead>
          <tbody>
            {currentItems.map((entry, index) => (
              <tr key={index}>
                <td>{entry.fileName}</td>
                <td>{entry.author}</td>
                <td>{entry.date}</td>
                <td>{entry.size}</td>
              </tr>
            ))}
          </tbody>
        </table>
        <ReactPaginate
          previousLabel={"Previous"}
          nextLabel={"Next"}
          pageCount={Math.ceil(mockConversionHistoryData.length / itemsPerPage)}
          onPageChange={changePage}
          containerClassName={"pagination"}
          previousLinkClassName={"pagination__link"}
          nextLinkClassName={"pagination__link"}
          activeClassName={"pagination__link--active"}
          disabledClassName={"pagination__link--disabled"}
        />
      </div>
    </div>
  );
};

export default ConversionHistory;
