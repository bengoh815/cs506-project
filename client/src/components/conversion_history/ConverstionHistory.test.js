import { cleanup, render, screen, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import ConversionHistory from "./ConversionHistory";

afterEach(() => {
    cleanup();
});

test("renders the ConversionHistory component", () => {
    render(<ConversionHistory />);

    // Use a more flexible text matcher to find the heading element
    expect(screen.getByRole("heading", { name: /conversion history/i })).toBeInTheDocument();
});

// Snapshot test to ensure UI changes are deliberate.
test("ConversionHistory component matches snapshot", () => {
    const tree = renderer.create(<ConversionHistory />).toJSON();

    expect(tree).toMatchSnapshot();
});

// Test file name sorting functionality
test("sorts conversion history by file name", () => {
    render(<ConversionHistory />);

    const fileNameColumn = screen.getByRole("columnheader", { name: /file/i });
    fireEvent.click(fileNameColumn);

    expect(screen.getByText("song9.mid")).toBeInTheDocument();
});


// Test file name sorting functionality
test("sorts conversion history by author name", () => {
    render(<ConversionHistory />);

    const authorNameColumn = screen.getByRole("columnheader", { name: /author/i });
    fireEvent.click(authorNameColumn);

    expect(screen.getByText("Alice")).toBeInTheDocument();
});


// Test date name sorting functionality
test("sorts conversion history by author name", () => {
    render(<ConversionHistory />);

    const dateColumn = screen.getByRole("columnheader", { name: /date/i });
    fireEvent.click(dateColumn);

    expect(screen.getByText("2024-03-10")).toBeInTheDocument();
});

// Test date name sorting functionality
test("sorts conversion history by author name", () => {
    render(<ConversionHistory />);

    const sizeColumn = screen.getByRole("columnheader", { name: /size/i });
    fireEvent.click(sizeColumn);

    expect(screen.getByText("0.9 MB")).toBeInTheDocument();
});

// Test pagination initial rows
test("displays correct number of items per page", () => {
    render(<ConversionHistory />);

    const items = screen.getAllByRole("row");
    expect(items.length).toBe(11);
});

// Test pagination switch button
test("displays correct items when changing pages", () => {
    render(<ConversionHistory />);

    // Attempt to find the "Next" button using a case-insensitive regex for the label
    const nextPageButton = screen.getByRole("button", { name: /2/i });
    fireEvent.click(nextPageButton);

    // After clicking next page, check if the correct items are displayed
    const itemsSecondPage = screen.getAllByRole("row");
    expect(itemsSecondPage.length).toBe(4); 
});

// Test the button number
test("counts the number of pages in the table", () => {
    render(<ConversionHistory />);

    const pageButtons = screen.getAllByRole("button", { name: /page/i });

    // Check if the number of page buttons matches the expected number of pages
    expect(pageButtons.length).toBe(4);
});