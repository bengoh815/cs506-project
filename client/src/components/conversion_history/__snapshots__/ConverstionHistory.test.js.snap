// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConversionHistory component matches snapshot 1`] = `
<div
  className="conversion-history-container"
>
  <div
    id="conversion-history"
  >
    <h2
      id="conversion-history-heading"
    >
      Conversion History
    </h2>
    <table
      id="conversion-history-table"
    >
      <thead>
        <tr>
          <th
            className="sortable"
            onClick={[Function]}
          >
            File
             â–²
          </th>
          <th
            className="sortable"
            onClick={[Function]}
          >
            Author
          </th>
          <th
            className="sortable"
            onClick={[Function]}
          >
            Date
          </th>
          <th
            className="sortable"
            onClick={[Function]}
          >
            Size
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            song1.mid
          </td>
          <td>
            Fiona
          </td>
          <td>
            2024-03-10
          </td>
          <td>
            3.4 MB
          </td>
        </tr>
        <tr>
          <td>
            song10.mid
          </td>
          <td>
            George
          </td>
          <td>
            2024-03-19
          </td>
          <td>
            1.4 MB
          </td>
        </tr>
        <tr>
          <td>
            song11.mid
          </td>
          <td>
            Diana
          </td>
          <td>
            2024-03-20
          </td>
          <td>
            254 MB
          </td>
        </tr>
        <tr>
          <td>
            song12.mid
          </td>
          <td>
            Alice
          </td>
          <td>
            2024-03-21
          </td>
          <td>
            3.0 MB
          </td>
        </tr>
        <tr>
          <td>
            song13.mid
          </td>
          <td>
            Bob
          </td>
          <td>
            2024-03-22
          </td>
          <td>
            0.9 MB
          </td>
        </tr>
        <tr>
          <td>
            song2.mid
          </td>
          <td>
            George
          </td>
          <td>
            2024-03-11
          </td>
          <td>
            3.5 MB
          </td>
        </tr>
        <tr>
          <td>
            song3.mid
          </td>
          <td>
            George
          </td>
          <td>
            2024-03-12
          </td>
          <td>
            5.4 MB
          </td>
        </tr>
        <tr>
          <td>
            song4.mid
          </td>
          <td>
            Hannah
          </td>
          <td>
            2024-03-13
          </td>
          <td>
            35.41 MB
          </td>
        </tr>
        <tr>
          <td>
            song5.mid
          </td>
          <td>
            Julia
          </td>
          <td>
            2024-03-14
          </td>
          <td>
            9.4 MB
          </td>
        </tr>
        <tr>
          <td>
            song6.mid
          </td>
          <td>
            Ethan
          </td>
          <td>
            2024-03-15
          </td>
          <td>
            13.0 MB
          </td>
        </tr>
      </tbody>
    </table>
    <ul
      aria-label="Pagination"
      className="pagination"
      role="navigation"
    >
      <li
        className="previous pagination__link--disabled"
      >
        <a
          aria-disabled="true"
          aria-label="Previous page"
          className="pagination__link "
          onClick={[Function]}
          onKeyPress={[Function]}
          rel="prev"
          role="button"
          tabIndex="-1"
        >
          Previous
        </a>
      </li>
      <li
        className="pagination__link--active"
      >
        <a
          aria-current="page"
          aria-label="Page 1 is your current page"
          onClick={[Function]}
          onKeyPress={[Function]}
          rel="canonical"
          role="button"
          tabIndex="-1"
        >
          1
        </a>
      </li>
      <li>
        <a
          aria-current={null}
          aria-label="Page 2"
          onClick={[Function]}
          onKeyPress={[Function]}
          rel="next"
          role="button"
          tabIndex="0"
        >
          2
        </a>
      </li>
      <li
        className="next"
      >
        <a
          aria-disabled="false"
          aria-label="Next page"
          className="pagination__link"
          onClick={[Function]}
          onKeyPress={[Function]}
          rel="next"
          role="button"
          tabIndex="0"
        >
          Next
        </a>
      </li>
    </ul>
  </div>
</div>
`;
